cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java

```diff
@@ -202,4 +202,72 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
     }
   }
 
+  public void restartServiceInstances(HttpRequest request, HttpResponder responder, final String serviceName) {
+    long startTimeMs = System.currentTimeMillis();
+    boolean isSuccess = true;
+    try {
+      if (!serviceManagementMap.containsKey(serviceName)) {
+        responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Invalid service name %s", serviceName));
+        return;
+      }
+
+      // Need to get Twill controller for the service.
+      MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
+      if (masterServiceManager.restartAllInstances()) {
+        responder.sendStatus(HttpResponseStatus.OK);
+      } else {
+        isSuccess = false;
+        responder.sendString(HttpResponseStatus.BAD_REQUEST, "Operation did not succeed");
```

Comment:
Shouldn't this actually be Internal Server Error rather than Bad Request? Bad Request is more for something wrong with the input/request.